snafu-derive/src/lib.rs

```diff
@@ -203,17 +204,24 @@ impl fmt::Display for DuplicateAttribute {
 /// it's added to a list of DuplicateAttribute errors, using the given `name` and `location` as
 /// descriptors.
 ///
-/// When done parsing a structure, call `finish`; it gives you the first attribute found, if any,
-/// and the list of errors.
+/// When done parsing a structure, call `finish` to get first attribute found, if any, and the list
+/// of errors, or call `finish_with_location` to get the attribute and the token tree where it was
```

Comment:
I'm surprised that there are times where we want to call just `finish` and *not* care about the error location...

Reply:
I thought so too, and originally I just had `finish_with_location`, but I found that throwing away the extra field in the tuple was unergonomic in some cases, and it seemed like minimal overhead to have a function for each case.  There are 6 uses of `finish` which felt sufficient to me.  But if it stood out to you too, maybe I should try again?

Reply:
I thought so too, and originally I just had `finish_with_location`, but I found that throwing away the extra field in the tuple was unergonomic in some cases, and it seemed like minimal overhead to have a function for each case. There are 6 uses of `finish` which felt like a good case to me in the end. But if this stood out to you too, maybe I should try again?

Reply:
> 6 uses of `finish` which felt like a justification to me in the end

Yes, so long as there are cases where the information isn't needed / used, it makes sense. My question is mostly around *why* isn't it needed in those cases, why was it being thrown away?

Reply:
> My question is mostly around why isn't it needed in those cases, why was it being thrown away?

Most error handling is done inside `AtMostOne` now, and the errors are surfaced when `finish` is called so they can be added to a `SyntaxErrors` or returned directly.  The location is useful for manual error checks, when we want to check for a condition other than there being more than one of an attribute, for example checking whether two incompatible attributes are used.  In those cases, you need the location to manually add an error.  In most cases, though, the location handling is done when `add`ing things to an `AtMostOne`.

An alternative could be to abstract `AtMostOne` into a more general error-handling structure that can check for duplication and other types of errors like incompatibility, but to me it seems like that would be quite a bit of complexity to handle the relatively few non-duplication errors we have now.  (I'm not counting attribute misuse errors, because in those cases you can immediately add an error to the list and don't need any higher-level tracking.)